### PR TITLE
Action mask prototype

### PR DESCRIPTION
**Pull request by ChengYen-Tang**

### CODE REVIEW

File: stable_baselines/common/policies.py [hunk 1]
```diff
@@ -122,6 +122,13 @@ def __init__(self, sess, ob_space, ac_space, n_env, n_steps, n_batch, reuse=Fals
             if add_action_ph:
                 self._action_ph = tf.placeholder(dtype=ac_space.dtype, shape=(n_batch,) + ac_space.shape,
                                                  name="action_ph")
+            if isinstance(ac_space, MultiDiscrete):
+                self._action_mask_ph = tf.placeholder(dtype=tf.float32, shape=(n_env, sum(ac_space.nvec)), name="action_mask_ph")
+            elif isinstance(ac_space, Discrete) or isinstance(ac_space, MultiBinary):
+                self._action_mask_ph = tf.placeholder(dtype=tf.float32, shape=(n_env, ac_space.n), name="action_mask_ph")
+            elif isinstance(ac_space, Box):
+                self._action_mask_ph = tf.placeholder(dtype=tf.float32, shape=(n_env, ac_space.shape[0]), name="action_mask_ph")
```
Comment (ChengYen-Tang) [inline stable_baselines/common/policies.py]: if i have 4 environment and action space is MultiDiscrete([2,3])
self._action_mask_ph shape=(4, 5)

File: stable_baselines/common/distributions.py [hunk 2]
```diff
@@ -335,15 +351,16 @@ def fromflat(cls, flat):
...
         """
         Probability distributions from multicategorical input
 
         :param nvec: ([int]) the sizes of the different categorical inputs
         :param flat: ([float]) the categorical logits input
         """
         self.flat = flat
-        self.categoricals = list(map(CategoricalProbabilityDistribution, tf.split(flat, nvec, axis=-1)))
+        self.action_mask = action_mask
+        self.categoricals = list(map(CategoricalProbabilityDistribution, tf.split(flat, nvec, axis=-1), tf.split(action_mask, nvec, axis=-1)))
```
Comment (ChengYen-Tang) [inline stable_baselines/common/distributions.py]: The shape of the flat here will be (4, 5)
nvec is [2, 3]
action_mask shape=(4, 5)

If the flat data is now [[1, 2, -3, -4, 5], [-6, 7, -8, 9, 10], [1, 2, -3, -4, 5], [-6, 7, -8, 9, 10]], after tf.split(flat, nvec, axis=-1), the data becomes [[[1,2], [-6, 7], [1,2] , [-6, 7]], [[-3, -4, 5], [-8, 9, 10], [-3, -4, 5], [-8, 9, 10]]]

The same principle, the action mask data is [[-inf, 0, -inf, 0, -inf], [0, -inf, 0, -inf, 0], [-inf, 0, 0, 0, -inf ], [ 0, -inf, -inf, -inf, 0]], after tf.split(action_mask, nvec, axis=-1), the data becomes [[[-inf, 0], [0, -inf ], [-inf, 0], [0, -inf]], [[-inf, 0, -inf], [0, -inf, 0], [0, 0, -inf] , [-inf, -inf, 0]]]

File: stable_baselines/common/distributions.py [hunk 3]
```diff
@@ -278,24 +278,31 @@ def sample_dtype(self):
...
 class CategoricalProbabilityDistribution(ProbabilityDistribution):
-    def __init__(self, logits):
+    def __init__(self, logits, action_mask):
         """
         Probability distributions from categorical input
 
         :param logits: ([float]) the categorical logits input
         """
         self.logits = logits
+        self.action_mask = action_mask
```
Comment (ChengYen-Tang) [inline stable_baselines/common/distributions.py]: Now I will receive one of the data.
logits = [[1, 2]，[-6, 7]，[1, 2]，[-6, 7]]
action_mask = [[-inf, 0], [0, -inf ], [-inf, 0], [0, -inf]]

File: stable_baselines/common/distributions.py [hunk 4]
```diff
@@ -293,11 +293,9 @@ def flatparam(self):
...
         # mask: 0 is valid action, -inf is invalid action
         # [1, 2, 3] add [0, -inf, 0] = [1, -inf, 3]
-        if np.shape(self.action_mask) == np.shape(self.logits):
-            logits = tf.add(self.logits, self.action_mask)
-        else:
-            logits = self.logits
-        return tf.argmax(self.logits, axis=-1)
+        logits = self.logits
+        logits = tf.add(logits, self.action_mask)
+        return tf.argmax(logits, axis=-1)
```
Comment (ChengYen-Tang) [inline stable_baselines/common/distributions.py]: logits = [[1, 2]，[-6, 7]，[1, 2]，[-6, 7]]
action_mask = [[-inf, 0], [0, -inf ], [-inf, 0], [0, -inf]]
[[1, 2]，[-6, 7]，[1, 2]，[-6, 7]] add [[-inf, 0], [0, -inf ], [-inf, 0], [0, -inf]] = [[-inf, 2], [-6, -inf], [-inf, 2], [-6, -inf]]

File: stable_baselines/common/policies.py [hunk 5]
```diff
@@ -214,12 +222,12 @@ def proba_step(self, obs, state=None, mask=None):
...
-        if np.shape(self.action_mask_ph) == np.shape(action_mask):
+        if np.shape(self.action_mask_shape) == np.shape(action_mask):
             action_mask = np.array(action_mask, dtype=np.float32)
             action_mask[action_mask <= 0] = -np.inf
             action_mask[action_mask > 0] = 0
         else:
-            action_mask = np.zeros(np.shape(self.action_mask_ph))
+            action_mask = np.zeros(np.shape(self.action_mask_shape))
 
         return action_mask
```
Comment (ChengYen-Tang) [inline stable_baselines/common/policies.py]: Action masks are only needed when interacting with the environment, train_model does not need.
